### ISSUE TITLE
Add language-specific code blocks

### ISSUE THREAD

**Issue by Nava2** [umple/umple#255]
Originally reported on Google Code with ID 356
Owned by `redeyedmars`

---

It would be very useful in many contexts to be able to specify the target language for
a block of code.

Currently we have blocks of code in, at least:
- before/after codeInjections
- state machine guards
- state machine actions
- statemachine activities
- method bodies

The notation is simply, for example

before setName {doSomething};

The requirement here is to allow, alternatively one or more language-specific versions
of the code, using a syntax something like the following:

before setName Java,Cpp.Php {doSomething();} Ruby {doSomething}

If a no language is specified, things would work as they do now.  

There could also be a default for the version not preceded by a language.

The code generator would use the language-specific version if there is one, or else
the default.

It is suggested that this be applied to before/after statements first. That would require
adding additional variables to the CodeInjection block. Since we are not planning on
adding too many languages, I think it is safe to just add variables javaCode, phpCode,
cppCode, rubyCode.

The grammar could just be enhanced to be something like the following:

beforeCode : before [operationName] [[codeToInject]]([[codeToInject ]])*
afterCode : after [operationName] [[codeToInject]]([[codeToInject ]])*
codeToInject- : ([[codelang]](, [[codelang]])\* )? { [**code] }
codelang : Java | Php | Ruby | Cpp

---

Reported by @umple on 2012-11-24 00:43:57

**Comment by Nava2** [comment]
Reported by @umple on 2012-12-31 17:41:55
- **Status changed**: `Accepted`
- **Labels added**: Diffic-Med, ucosp
- **Labels removed**: Diffic-Easy

**Comment by Nava2** [comment]
Reported by @umple on 2013-01-19 16:51:41
- **cc added:** +`redeyedmars`

**Comment by Nava2** [comment]
Reported by @umple on 2013-01-19 17:12:42
- **Owner changed**: `redeyedmars`
- **cc removed:** -`redeyedmars`

**Comment by Nava2** [comment]
The grammar I used to fix this change is pretty bad, but the implementation works to
the extent that I was able to test it. I have given up for now with fighting with the
grammar because I don't know how to debug it, if I try one style that should work,
it spits out a random null pointer exception, very strange. So if anyone else can fix
it, I would very much like to see how, otherwise I shall come back later and try to
fix it then.

---

Reported by `redeyedmars` on 2013-01-21 00:41:32
